src/ripple/ledger/impl/ApplyView.cpp

```diff
@@ -0,0 +1,298 @@
+    }
+
+    // The current page is now empty; check if it can be
+    // deleted, and, if so, whether the entire directory
+    // can now be removed.
+    auto prevPage = node->getFieldU64(sfIndexPrevious);
+    auto nextPage = node->getFieldU64(sfIndexNext);
+
+    // The first page is the directory's root node and is
+    // treated specially: it can never be deleted even if
+    // it is empty, unless we plan on removing the entire
+    // directory.
+    if (currPage == 0)
+    {
+        if (nextPage == currPage && prevPage != currPage)
+            LogicError ("Directory chain: fwd link broken");
+
+        if (prevPage == currPage && nextPage != currPage)
+            LogicError ("Directory chain: rev link broken");
+
+        // Older versions of the code would, in some cases,
+        // allow the last page to be empty. Remove such
+        // pages if we stumble on them:
+        if (nextPage == prevPage && nextPage != currPage)
+        {
+            auto last = peek(keylet::page(directory, nextPage));
+            if (!last)
+                LogicError ("Directory chain: fwd link broken.");
+
+            if (last->getFieldV256 (sfIndexes).empty())
+            {
+                // Update the first page's linked list and
+                // mark it as updated.
+                node->setFieldU64 (sfIndexNext, currPage);
+                node->setFieldU64 (sfIndexPrevious, currPage);
+                update(node);
+
+                // And erase the empty last page:
+                erase(last);
+
+                // Make sure our local values reflect the
+                // updated information:
+                nextPage = currPage;
+                prevPage = currPage;
+            }
+        }
+
+        if (keepRoot)
+            return true;
+
+        // If there's no other pages, erase the root:
+        if (nextPage == currPage && prevPage == currPage)
+            erase(node);
+
+        return true;
+    }
+
+    // This can never happen for nodes other than the root:
+    if (nextPage == currPage)
+        LogicError ("Directory chain: fwd link broken");
+
+    if (prevPage == currPage)
+        LogicError ("Directory chain: rev link broken");
+
+    // This node isn't the root, so it can either be in the
+    // middle of the list, or at the end. Unlink it first
+    // and then check if that leaves the list with only a
+    // root:
+
+    auto prev = peek(keylet::page(directory, prevPage));
+    if (!prev)
+        LogicError ("Directory chain: fwd link broken.");
+
+    // Fix previous to point to its new next.
+    prev->setFieldU64(sfIndexNext, nextPage);
+    update (prev);
+
+    auto next = peek(keylet::page(directory, nextPage));
+    if (!next)
+        LogicError ("Directory chain: rev link broken.");
+
+    // Fix next to point to its new previous.
+    next->setFieldU64(sfIndexPrevious, prevPage);
+    update(next);
+
+    // The page is no longer linked. Delete it.
+    erase(node);
```

Comment:
At this point we need to check if there are two pages left and the last page is empty. If we don't, we won't handle the [EMPTY] -> [Just removed last element] -> [EMPTY] scenario. If guessing the unit tests pass because the last page is no longer allowed to be empty. It it were (i.e. an existing ledger), then I would expect the unit test to fail.